### PR TITLE
upgrade std.string.column to handle ranges and nelSep

### PR DESCRIPTION
**Pull request by WalterBright**
String functions should handle input ranges of text. Also upgraded the unit tests for `column()` so the coverage is complete.

### CODE REVIEW

File: std/string.d [hunk 1]
```diff
@@ -4722,46 +4722,76 @@ string[string] abbrev(string[] values) @safe pure
...
  * Returns:
  *    column number
  */
 
-size_t column(S)(S str, in size_t tabsize = 8) @safe pure @nogc if (isSomeString!S)
+size_t column(Range)(Range str, in size_t tabsize = 8)
+    if (!isStaticArray!Range &&
+        (is(Range : const  char[]) ||
+         is(Range : const wchar[]) ||
+         is(Range : const dchar[]) ||
```
Comment (burner) [inline std/string.d]: shouldn't this be $(D isSomeString!Range)

File: std/string.d [hunk 2]
```diff
@@ -4722,46 +4722,76 @@ string[string] abbrev(string[] values) @safe pure
...
+         isInputRange!Range &&
+             (is(Unqual!(ElementEncodingType!Range) ==  char) ||
+              is(Unqual!(ElementEncodingType!Range) == wchar) ||
+              is(Unqual!(ElementEncodingType!Range) == dchar))
```
Comment (burner) [inline std/string.d]: shouldn't this be $(D isSomeChar!(Unqual!(ElementEncodingType!Range)) )

File: std/string.d [hunk 3]
```diff
@@ -4722,46 +4722,76 @@ string[string] abbrev(string[] values) @safe pure
...
 {
+    import std.utf : byChar, byWchar, byDchar;
+
     assert(column("1234 ") == 5);
+    assert(column("1234 "w) == 5);
+    assert(column("1234 "d) == 5);
+
+    assert(column("1234 ".byChar()) == 5);
+    assert(column("1234 "w.byWchar()) == 5);
+    assert(column("1234 "d.byDchar()) == 5);
```
Comment (burner) [inline std/string.d]: Could you add tests with all the DummyRanges

File: std/string.d [hunk 4]
```diff
@@ -4722,46 +4722,76 @@ string[string] abbrev(string[] values) @safe pure
...
+    }
+    else
     {
-        switch (c)
+        // decoding not needed for wchars and dchars
+        import std.uni : lineSep, paraSep, nelSep;
+
+        size_t column;
+
+        foreach (const c; str)
```
Comment (JakobOvrum) [inline std/string.d]: Why `const`? Surely `immutable` is a more appropriate cue?

File: std/string.d (hunk 3 above)
Comment (WalterBright) [inline std/string.d]: Not sure what you mean. The .byChar, .byWchar, and .byDchar does test column() with range input.

File: std/string.d (hunk 4 above)
Comment (WalterBright) [inline std/string.d]: const for variables is effectively immutable, and also works when the value being assigned is not immutable.

File: std/string.d (hunk 1 above)
Comment (WalterBright) [inline std/string.d]: yes

File: std/string.d (hunk 2 above)
Comment (WalterBright) [inline std/string.d]: yes

File: std/string.d (hunk 3 above)
Comment (burner) [inline std/string.d]: yes, but there are many more types of ranges other than by(W|D)Char. what about infinite ranges, char[] etc. as I understand the Dummy Ranges. They give you Input Ranges, Forward Ranges, Random Access, and Infinite Ranges. Column will accept everything that is at least an Input Range, so I would like to see those tested

Comment (WalterBright) [inline std/string.d]: They won't be tested anyway. column() only uses the most basic features of an InputRange (what `foreach` does).

Comment (burner) [inline std/string.d]: I think you're right. The only use I see is the foreach.

**Comment by burner** [comment]
LGTM

**Comment by burner** [comment]
Auto-merge toggled on

**Pull request merged**